onnxruntime/core/providers/mkldnn/nn/conv.cc

```diff
@@ -401,23 +401,35 @@ Status Conv<T>::Compute(OpKernelContext* context) const {
       DoReorder<T>(params);
       src_data = static_cast<T*>(dst.get_data_handle());
     }
-
-    // Reorder filter memory layout if necessary.
-    if (filter_format != conv_primitive->GetFilterMemoryFormat()) {
-      auto pd = mkldnn::memory::primitive_desc(mkldnn::memory::desc(filter_dims_mkl,
-                                                                    MklDnnType<T>(),
-                                                                    filter_format),
-                                               cpu_engine);
-      mkldnn::memory src = mkldnn::memory(pd, (void*)filter_data);
-      // allocate the size queried from memory primitive desc. it may not match tensor logical size due to
-      // mkldnn using padding to allow use of blocked format.
-      filter_reorder_buffer = IAllocator::MakeUniquePtr<void>(alloc, conv_primitive->GetFilterSize());
-      mkldnn::memory dst = mkldnn::memory(conv_fwd_pd->weights_primitive_desc(), filter_reorder_buffer.get());
-      MemoryReorderParams params(src, dst);
-      DoReorder<T>(params);
-      filter_data = static_cast<T*>(dst.get_data_handle());
+    // Reorder filter memory layout if necessary
+    // Avoid data reordering. Save filter memory in mkldnn format from first iteration 
+    // in execution provider mapped by weight name.
+    std::string weightKey = OpKernel::Node().InputDefs()[1]->Name();
+    std::shared_ptr<mkldnn::memory> filter_dst_mem =nullptr;
+    filter_dst_mem = provider_->GetWeightMemory(weightKey);
+
+    if (filter_dst_mem == nullptr) {
+      if (filter_format != conv_primitive->GetFilterMemoryFormat()) {
+        auto pd = mkldnn::memory::primitive_desc(mkldnn::memory::desc(
+          filter_dims_mkl, MklDnnType<T>(), filter_format), cpu_engine);
+        mkldnn::memory src = mkldnn::memory(pd, (void*)filter_data);
+        filter_reorder_buffer = IAllocator::MakeUniquePtr<void>(alloc, conv_primitive->GetFilterSize());
+        filter_dst_mem.reset(
+          new mkldnn::memory(conv_fwd_pd->weights_primitive_desc(), filter_reorder_buffer.get()));
+        MemoryReorderParams params(src, *filter_dst_mem);
+        DoReorder<T>(params);
+        filter_data = static_cast<T*>(filter_dst_mem->get_data_handle());
+        {
+          // make assignment threadsafe
+          std::lock_guard<std::mutex> lock(mutex_);
```

Comment:
@jywu-msft will move lock to SetWeightMemory() of provider.

Re Atomic: are you suggesting to have one thread do the atomic operation of "checking + Reorder + Setting of all weights"?

Reply:
we only want 1 thread to do the weights reordering.
with your existing code, multiple threads could do the reordering since you are only synchronizing the setting of the map.
alternatively, maybe you can try using std::call_once() instead of using a mutex.

Reply:
@jywu-msft can you please take a look?

created new function Reorder_weights in Conv (too many parameters) so that I can call std::call_once but couldn't figure out how to pass parameters to call_once args. 

Created Get_weights_memory and Set_weights_memory and moved lock and mutex to provider (for now, Will remove after call_once implementation)

Reply:
did you test these changes? i saw failures in Conv unit tests when building these changes.
re: std::call_once, it would be something like

std::call_once(flag1, &Conv::Reorder_weights, this, context, src_dims_mkl, filter_dims_mkl, bias_dims_mkl, dst_dims_mkl, strides_mkl, dilations_mkl, padding_left_mkl, padding_right_mkl, filter_format, filter_dims_mkl);

Reply:
@jywu-msft 

`std::call_once` didn't work and it's called once in the process. Only first test passed and rest of the tests failed. `perf_test `(Resnet50) also fails.

I've moved the call_once implementation to new [branch](https://github.com/sreekanth-yalachigere/onnxruntime/blob/f2bf757d972d0ded02fbdd5f57bd5fe0762e3c66/onnxruntime/core/providers/mkldnn/nn/conv.cc#L435), . 

Here, I am using mutex for thread-safe implementation. 

```
check memory map
if (empty)
lock(){
  check memory map // did last thread re-order and save?
  if (empty){
     re-order
     save in weight-name and memory map
  }
}
```

Reply:
lock should be owned by the provider, as the map is also owned by provider. 
as currently implemented, you can still have multiple threads check the memory map and see it empty (and then proceed to do the reorder)
the check memory map needs to be done while holding lock. 
please ensure you build with --enable_onnx_tests and tested on onnx models. 
seems like a bunch of models are failing in the CI run.